### PR TITLE
Wrap up data storing classes

### PR DESCRIPTION
**Pull request by ldcWV**

### CODE REVIEW

File: src/main/java/com/google/sps/data/Game.java
Comment (saramissak) [inline src/main/java/com/google/sps/data/Game.java]: Try reading this out loud, it sounds off.

File: src/main/java/com/google/sps/data/Game.java
Comment (saramissak) [inline src/main/java/com/google/sps/data/Game.java]: Rather than putting val, make the variable names more clear what they are for. I understand that there is only one variable so it doesn't make a difference, but for the API when people hover over the function and look at the parameters, having a more specific name will help since they won't see the implementation of it.

File: src/main/java/com/google/sps/data/Game.java
Comment (saramissak) [inline src/main/java/com/google/sps/data/Game.java]: for this Game.*

File: src/main/java/com/google/sps/data/Game.java
Comment (saramissak) [inline src/main/java/com/google/sps/data/Game.java]: Maybe change this and the other ones from "a String representation" to "a String containing" 
Having it as "representation" makes me think that you are saying it was some other type and then you are changing the type in the function.

File: src/main/java/com/google/sps/data/Pair.java
Comment (saramissak) [inline src/main/java/com/google/sps/data/Pair.java]: What's T1 and T2? What are these pairs for? Could you be more specific here I am confused.

File: src/main/java/com/google/sps/data/Progress.java
Comment (saramissak) [inline src/main/java/com/google/sps/data/Progress.java]: As I was talking before with the other comment about changing the val to be more specific.

**Review by saramissak — APPROVED** [review]

File: src/main/java/com/google/sps/data/Game.java
Comment (ldcWV) [reply src/main/java/com/google/sps/data/Game.java]: Makes sense, I've done this for every time I had val. Thanks!

File: src/main/java/com/google/sps/data/Game.java
Comment (ldcWV) [reply src/main/java/com/google/sps/data/Game.java]: Reworded as "the number of times the game has received star votes", thanks.

File: src/main/java/com/google/sps/data/Game.java
Comment (ldcWV) [reply src/main/java/com/google/sps/data/Game.java]: Done.

File: src/main/java/com/google/sps/data/Game.java
Comment (ldcWV) [reply src/main/java/com/google/sps/data/Game.java]: I agree that sounds better, done.

File: src/main/java/com/google/sps/data/Pair.java
Comment (ldcWV) [reply src/main/java/com/google/sps/data/Pair.java]: T1 and T2 are arbitrary classes. The purpose of a pair is to store 2 objects together. This is used in the User class in order to store the games a user has played along with the amount of time they spent on them.

File: src/main/java/com/google/sps/data/Progress.java
Comment (ldcWV) [reply src/main/java/com/google/sps/data/Progress.java]: Done.

File: src/main/java/com/google/sps/data/Pair.java
Comment (saramissak) [reply src/main/java/com/google/sps/data/Pair.java]: Could you make it type1 and type2 rather than T1 and T2?

Comment (ldcWV) [reply src/main/java/com/google/sps/data/Pair.java]: Yep, done. Thanks!

File: src/main/java/com/google/sps/data/Coordinates.java
Comment (shenqiuz) [inline src/main/java/com/google/sps/data/Coordinates.java:54]: Nice! We should avoid those "val", "temp" naming in the code.

File: src/main/java/com/google/sps/data/Game.java
Comment (shenqiuz) [inline src/main/java/com/google/sps/data/Game.java:188]: The "this" is only needed when the local variable and the member variable have the same name. Or the member variable (i.e. this.value) and the parent variable (i.e. super.value) have the same name. Except for those cases, we don't need to put "this" here.

public String getGameID() {
     return gameID;
}

File: src/main/java/com/google/sps/data/Pair.java
Comment (shenqiuz) [inline src/main/java/com/google/sps/data/Pair.java:22]: Can the "Map.Entry" service the purpose?
https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html

File: src/main/java/com/google/sps/data/User.java
Comment (shenqiuz) [inline src/main/java/com/google/sps/data/User.java:28]: For "gamesCreated", I expected to have a List<Game>. Since we have List<String>, I think it might be more clear to have List<String> gameIDsCreated.

File: src/main/java/com/google/sps/data/User.java
Comment (shenqiuz) [inline src/main/java/com/google/sps/data/User.java]: As for "String profilePicture", should we use "profilePictureUrl"? If it is a profile picture, then I expect the type should be a picture type.

File: src/main/java/com/google/sps/data/User.java
Comment (shenqiuz) [inline src/main/java/com/google/sps/data/User.java]: You don't need to initialize "username", "firstName", "lastName", "profilePicture" to null. The following code does the same thing by initializing them to null.

private String username;
private String firstName;
private String lastName;
private String profilePicture;

File: src/main/java/com/google/sps/data/User.java
Comment (shenqiuz) [inline src/main/java/com/google/sps/data/User.java]: For the List, sometimes it might crash by initializing it to null. For example, it is quite comment that some other function might check the size of the list (i.e. gamesCreated.size()). In this case, setting gamesCreated to null will crash the code. So a better initial value could be an empty list.

private List<String> gamesCreated = new ArrayList<>();

File: src/main/java/com/google/sps/data/User.java
Comment (shenqiuz) [inline src/main/java/com/google/sps/data/User.java]: Why don't you use a Map?

File: src/main/java/com/google/sps/servlets/LoadGameServlet.java
Comment (shenqiuz) [inline src/main/java/com/google/sps/servlets/LoadGameServlet.java]: For those hardcoded mock username, please consider to use constants and put all those constants at the beginning of this java file. In this way, it is easy to deprecate those mock fields in the future. Otherwise, you might need to go through the code to find those mock variables.

public class LoadGameServlet extends HttpServlet {
   private static final String USER_NAME = "usernameid12345";
   ....
         gameBuilder.setGameCreator(USER_NAME);
}

File: src/main/java/com/google/sps/servlets/CreateGameServlet.java
Comment (shenqiuz) [inline src/main/java/com/google/sps/servlets/CreateGameServlet.java]: It is better to return an empty List.

File: src/main/java/com/google/sps/servlets/CreateGameServlet.java
Comment (shenqiuz) [inline src/main/java/com/google/sps/servlets/CreateGameServlet.java:161]: As for the "CreateGameServlet.java", those methods are correlated with each other. For example, the returning value of List<Integer> getNumHints should match the number of elements for each row in List<List<Coordinates>> getHintLocations(). What would happen if the returning values of both methods don't agree with each other. It is aways better to have a single-source-of-truth. For example, we might have some methods like below. I didn't put too much thoughts when listing the following methods. Please use your judgment when create those methods. 

private int getNumStages();

private int getNumHints(int stageIndex);

private List<Coordinates> getHintLocations(int stageIndex);

private Coordinate getHintLocation(int stageIndex, int hintIndex);

File: src/main/java/com/google/sps/data/Game.java
Comment (ldcWV) [reply src/main/java/com/google/sps/data/Game.java:188]: I think this just makes the code slightly less ambiguous and doesn't have any drawbacks.

File: src/main/java/com/google/sps/data/User.java
Comment (ldcWV) [reply src/main/java/com/google/sps/data/User.java]: This allows us to store two or more different times for the same game. Also, I don't think we'll ever have to look up a time for a specific game.

File: src/main/java/com/google/sps/data/Pair.java
Comment (ldcWV) [reply src/main/java/com/google/sps/data/Pair.java:22]: Probably, but I think a Pair is clearer in that it doesn't necessarily consider the first element as a "key" and the second element as a "value". This sort of naming could be confusing if the relationship between the two elements is not like that. Also see https://github.com/googleinterns/step55-2020/pull/8#discussion_r447739305.

File: src/main/java/com/google/sps/data/User.java
Comment (ldcWV) [reply src/main/java/com/google/sps/data/User.java:28]: I don't understand your comment about the List. Do you mean to change ArrayList to List? I'm not sure why this is good because then it becomes unclear what type of List gamesCreated is. 

Also I don't think putting "ID" in the name of the variable is necessary because the default assumption over all the code is that when referencing a game, stage, hint, or user, we always use ID. Also this would require changing the names of a lot more variables, which would take a long time.

File: src/main/java/com/google/sps/data/User.java
Comment (ldcWV) [reply src/main/java/com/google/sps/data/User.java]: Yeah, I was thinking about this while coding it. I think I'll make this change, thanks!

File: src/main/java/com/google/sps/servlets/CreateGameServlet.java
Comment (ldcWV) [reply src/main/java/com/google/sps/servlets/CreateGameServlet.java]: Overwritten in the next commit.

File: src/main/java/com/google/sps/data/User.java
Comment (ldcWV) [reply src/main/java/com/google/sps/data/User.java]: I think in this case crashing is a good thing. If we have never explicitly set a value for the list, and someone is trying to access it anyways, then this should be a problem. Crashing is a good way to immediately point out that the list shouldn't be accessed. If it doesn't crash here and instead the client is interacting with an empty list, we would have to find out manually why the list is empty.

Setting the list to be initially empty should be something done explicitly outside the class.

File: src/main/java/com/google/sps/data/User.java
Comment (ldcWV) [reply src/main/java/com/google/sps/data/User.java]: I just like the style of this better because I'm setting all the ints to -1, so I would like to set the Strings to some value as well. I think looks weird to set some variables equal to some initial value, while leaving the others untouched, even though they get set to null by default.

File: src/main/java/com/google/sps/servlets/LoadGameServlet.java
Comment (ldcWV) [reply src/main/java/com/google/sps/servlets/LoadGameServlet.java]: Thanks, I will do this. I'm planning to move all this mock data into a mock data manager next, so I'll just set those variables there.

File: src/main/java/com/google/sps/servlets/CreateGameServlet.java
Comment (ldcWV) [reply src/main/java/com/google/sps/servlets/CreateGameServlet.java:161]: There's not much I can do here in regards to determining the single-source truth values, as these lists come directly from a parameters in a POST request. I've instead added a checker that makes sure the lists have sizes that agree, and will print an exception and terminate if not.

File: src/main/java/com/google/sps/data/Game.java
Comment (shenqiuz) [reply src/main/java/com/google/sps/data/Game.java:188]: Yeah, this could be a personal choice. The following link contains some information. As long as the whole team uses the same coding style, I think either way is acceptable. 

https://softwareengineering.stackexchange.com/questions/211319/excessive-use-this-keyword-in-java

File: src/main/java/com/google/sps/data/Pair.java
Comment (shenqiuz) [reply src/main/java/com/google/sps/data/Pair.java:22]: If you write your own Pair class, then you should always overwrite the equals and hashCode methods if you need to compare two Pair objects, or put the Pair objects into Set or Map. Because without overwriting the equals() and hashCode() methods, the equality checks the object-level equality. Please add an unit test like below for the Pair class.

Pair<String, String> pair1 = Pair("first", "second");
Pair<String, String> pair2 = Pair("first", "second");
assertThat(pair1).isEqualTo(pair2);

Without overwriting the equals() and hashCode(), then the above test might fail.

https://medium.com/codelog/overriding-hashcode-method-effective-java-notes-723c1fedf51c

File: src/main/java/com/google/sps/data/User.java
Comment (shenqiuz) [reply src/main/java/com/google/sps/data/User.java:28]: Yeah, you can use gamesCreated for now. Usually we use createdGameIds as the valuable name if the list stores a list of gameIds. Because we should review those variable names with the partner support engineers. Everything should be clear. But you can use gamesCreated in this project.

File: src/main/java/com/google/sps/data/User.java
Comment (shenqiuz) [reply src/main/java/com/google/sps/data/User.java]: Actually the code will crash at some random library rather than crash in your code. For example, suppose you use a library to print created game IDs. It will crash at "strings.size()". 

public String printListOfString(List<String> strings) {
   if (strings.size() == 0) {
      return "";
  }
 ... 
}

Also, we should consider the meaning of this class. This class is User.java and some users might not create any games. In this case, the gamesCreated should be initialized to an empty list.

File: src/main/java/com/google/sps/data/User.java
Comment (shenqiuz) [reply src/main/java/com/google/sps/data/User.java]: You can use Multimap (https://guava.dev/releases/23.0/api/docs/com/google/common/collect/Multimap.html).

File: src/main/java/com/google/sps/servlets/CreateGameServlet.java
Comment (shenqiuz) [reply src/main/java/com/google/sps/servlets/CreateGameServlet.java:161]: Why do you need to use the "HttpServletRequest request" as the argument for all private methods?

File: src/main/java/com/google/sps/data/Pair.java
Comment (ldcWV) [reply src/main/java/com/google/sps/data/Pair.java:22]: Got it, thanks!

File: src/main/java/com/google/sps/servlets/CreateGameServlet.java
Comment (ldcWV) [reply src/main/java/com/google/sps/servlets/CreateGameServlet.java:161]: The request contains all of the fetch parameters I need to parse information about the newly created game.

File: src/main/java/com/google/sps/data/User.java
Comment (ldcWV) [reply src/main/java/com/google/sps/data/User.java]: I just don't see the purpose of using a map here. Every time we access gamesCompletedWithTime, we're going to iterate through the entire list anyways. So the fast lookup speed of Map/Multimap, which is its main benefit, isn't going to be used at all.

File: src/main/java/com/google/sps/data/User.java
Comment (ldcWV) [reply src/main/java/com/google/sps/data/User.java]: Makes sense, I've made similar changes over all the files.

File: src/main/java/com/google/sps/servlets/LoadGameServlet.java
Comment (shenqiuz) [inline src/main/java/com/google/sps/servlets/LoadGameServlet.java:45]: You don't need to have a private method that is just used once and contains a single-line operation. It is much clear to parse the data from the request directly as below. 

String gameID = request.getParameter("gameID");

File: src/main/java/com/google/sps/servlets/CreateGameServlet.java
Comment (shenqiuz) [inline src/main/java/com/google/sps/servlets/CreateGameServlet.java:84]: I can see that the request contains a lot of fields, and do you have a doc of the defined fields? I think we should have an agreement on the JSON structure for the request and response before implementing it. I just checked the design doc and there is no JSON structure defined for the javascript-to-java interface. 

Also, it is hard to track down all these fields with so many private methods. You can have a single method to parse the HttpServletRequest like below:

private String userID;
private String gameDescription;

private void parseRequest(HttpServletRequest request) {
    userID = request.getParameter("userID");
    gameDescription = request.getParameter("gameDescription");
    ....
}

Then it is easier to see if all those fields make sense or not.

**Review by shenqiuz — COMMENTED** [review]
This pull request contains 73 files. Firstly it is hard to review such a big pull request. Secondly, it will take longer to resolve all comments and submit it. Thus, you should split the pull request to small ones. 

For example, the Game class and the User class are only used between Java and Datastore. You probably want to submit those classes to unblock Ade. In this case, you can send a small pull request with those two files. 

As for the CreateGameServlet, the JSON structure of the request and response are not written in the design doc, reviewed by all stakeholders. In this case, we might have a lot of back-and-forth discussions on those files.

File: src/main/java/com/google/sps/data/User.java
Comment (shenqiuz) [reply src/main/java/com/google/sps/data/User.java]: Do you plan to use the gamesCompletedWithTime to calculate the average time per game or the best play time per game? If this is the case, then you shouldn't store all raw <gameId, completedTime>. Because the list will keep growing and waste storage. Imagine that the wasted storage is huge if you have x10 million users. For the user statics purpose, try to aggregate the data and only store the necessary data (e.g. the average time or the best time for playing the game).

File: src/main/java/com/google/sps/servlets/CreateGameServlet.java
Comment (shenqiuz) [reply src/main/java/com/google/sps/servlets/CreateGameServlet.java:161]: Use a centralized method to parse the HttpServletRequest. So it is easy to check the JSON structure of the request and extend the JSON structure in the future..

File: src/main/java/com/google/sps/data/User.java
Comment (ldcWV) [reply src/main/java/com/google/sps/data/User.java]: The plan is to create a histogram with game difficulty on the x axis and time on the y axis. Also, I prefer storing all the raw data just in case we want to do something different with it later on. We don't expect this to be a problem because we don't expect to have 10 million users. :P

File: src/main/java/com/google/sps/servlets/CreateGameServlet.java
Comment (ldcWV) [reply src/main/java/com/google/sps/servlets/CreateGameServlet.java:161]: I have already listed out the full JSON structure needed in the Java API doc. The link to that doc can be found in the appendix of our design doc.

File: src/main/java/com/google/sps/servlets/LoadGameServlet.java
Comment (ldcWV) [reply src/main/java/com/google/sps/servlets/LoadGameServlet.java:45]: In my opinion, I think it's clearer to have a separate function for each individual parameter. This is because some parameters are formatted differently than others (Strings, ints, arrays, etc.), so I need to parse them in different ways. If I parse them all together in one function, each in its own way, things could get confusing.

File: src/main/java/com/google/sps/servlets/CreateGameServlet.java
Comment (ldcWV) [reply src/main/java/com/google/sps/servlets/CreateGameServlet.java:84]: Here's the doc with the defined fields. It was hidden in the appendix of our design doc, but I plan to add an APIs section to make them more visible: https://docs.google.com/document/d/1w0ovXFHzRoE2ai9Xa8TgWKkI9-l5IAxb9nI2xyH7rRo/edit

Regarding the parseRequest function, I do think this makes sense and I think I will do that. But I still prefer keeping all of my other parsing functions that I can call within parseRequest, just because they split the work up nicely.

**Comment by ldcWV** [comment]
> This pull request contains 73 files. Firstly it is hard to review such a big pull request. Secondly, it will take longer to resolve all comments and submit it. Thus, you should split the pull request to small ones.
> 
> For example, the Game class and the User class are only used between Java and Datastore. You probably want to submit those classes to unblock Ade. In this case, you can send a small pull request with those two files.
> 
> As for the CreateGameServlet, the JSON structure of the request and response are not written in the design doc, reviewed by all stakeholders. In this case, we might have a lot of back-and-forth discussions on those files.

Thanks for all your comments Shenqiu! I searched online and there doesn't seem to be an easy way to split up a branch by files, other than reverting the commits and manually staging parts of it at a time.

**Pull request merged**
